crates/matrix-sdk-base/src/read_receipts.rs

```diff
@@ -0,0 +1,396 @@
+use eyeball_im::Vector;
+use matrix_sdk_common::deserialized_responses::SyncTimelineEvent;
+use ruma::{
+    events::{
+        poll::{start::PollStartEventContent, unstable_start::UnstablePollStartEventContent},
+        receipt::{ReceiptThread, ReceiptType},
+        room::message::Relation,
+        AnySyncMessageLikeEvent, AnySyncTimelineEvent, OriginalSyncMessageLikeEvent,
+        SyncMessageLikeEvent,
+    },
+    serde::Raw,
+    EventId, OwnedEventId, RoomId, UserId,
+};
+use tracing::{instrument, trace};
+
+use super::BaseClient;
+use crate::{error::Result, store::StateChanges, RoomInfo};
+
+/// Provider for timeline events prior to the current sync.
+pub trait PreviousEventsProvider: Send + Sync {
+    /// Returns the list of known timeline events, in sync order, for the given
+    /// room.
+    fn for_room(&self, room_id: &RoomId) -> Vector<SyncTimelineEvent>;
+}
+
+impl PreviousEventsProvider for () {
+    fn for_room(&self, _: &RoomId) -> Vector<SyncTimelineEvent> {
+        Vector::new()
+    }
+}
+
+#[instrument(skip_all, fields(room_id = %room_info.room_id))]
+pub(crate) async fn compute_notifications(
+    client: &BaseClient,
+    changes: &StateChanges,
+    previous_events_provider: &dyn PreviousEventsProvider,
+    new_events: &[SyncTimelineEvent],
+    room_info: &mut RoomInfo,
+) -> Result<()> {
+    let user_id = &client.session_meta().unwrap().user_id;
+    let prev_latest_receipt_event_id = room_info.read_receipts.latest_read_receipt_event_id.clone();
+
+    if let Some(receipt_event) = changes.receipts.get(room_info.room_id()) {
+        trace!("Got a new receipt event!");
+
+        // Find a private or public read receipt for the current user.
+        let mut receipt_event_id = None;
+        if let Some((event_id, receipt)) =
+            receipt_event.user_receipt(user_id, ReceiptType::ReadPrivate)
+        {
+            if receipt.thread == ReceiptThread::Unthreaded || receipt.thread == ReceiptThread::Main
+            {
+                receipt_event_id = Some(event_id.to_owned());
+            }
+        } else if let Some((event_id, receipt)) =
+            receipt_event.user_receipt(user_id, ReceiptType::Read)
+        {
+            if receipt.thread == ReceiptThread::Unthreaded || receipt.thread == ReceiptThread::Main
+            {
+                receipt_event_id = Some(event_id.to_owned());
+            }
+        }
+
+        if let Some(receipt_event_id) = receipt_event_id {
+            // We've found the id of an event to which the receipt attaches. The associated
+            // event may either come from the new batch of events associated to
+            // this sync, or it may live in the past timeline events we know
+            // about.
+
+            // First, save the event id as the latest one that has a read receipt.
+            room_info.read_receipts.latest_read_receipt_event_id = Some(receipt_event_id.clone());
```

Comment:
This unconditionally saves a receipt we receive as the latest one. I don't know whether this can happen, but if we receive a receipt out of order, we may end up with incorrect unreadness.

Knowing which receipt is latest is particularly problematic, because we have to find the events they refer to, and then decide which of those is "latest".

Reply:
(I assume this can happen, and, to be explicit, I think the correct behaviour is to figure out which is latest and keep only that one.)

In [matrix-js-sdk](https://github.com/matrix-org/matrix-js-sdk/blob/develop/src/models/room-receipts.ts), we hold on to "dangling" receipts for which we don't have an event, so that when/if we get the event, we can use that to decide whether the receipt is after the one we already have.

Reply:
Very good point. Well it'll be for a later iteration of this PR I suppose :woozy_face: